google/cloud/bigtable/_read_rows.py

```diff
@@ -0,0 +1,667 @@
+        except Exception as e:
+            await buffer.put(e)
+
+    @staticmethod
+    async def _buffer_to_generator(
+        buffer: asyncio.Queue[Any],
+    ) -> AsyncGenerator[Any, None]:
+        """
+        Helper function to yield items from a buffer as an async generator
+        """
+        while True:
+            item = await buffer.get()
+            if item is StopAsyncIteration:
+                return
+            if isinstance(item, Exception):
+                raise item
+            yield item
+            await asyncio.sleep(0)
+
+    async def _read_rows_retryable_attempt(
+        self,
+        gapic_fn: Callable[..., Awaitable[AsyncIterable[ReadRowsResponse]]],
+        buffer_size: int,
+        per_row_timeout: float | None,
+        per_request_timeout: float | None,
+        total_row_limit: int,
+    ) -> AsyncGenerator[Row | RequestStats, None]:
+        """
+        Retryable wrapper for merge_rows. This function is called each time
+        a retry is attempted.
+
+        Some fresh state is created on each retry:
+          - grpc network stream
+          - buffer for the stream
+          - state machine to hold merge chunks received from stream
+        Some state is shared between retries:
+          - last_seen_row_key is used to ensure that
+            duplicate rows are not emitted
+          - request is stored and (optionally) modified on each retry
+        """
+        if self._last_seen_row_key is not None:
+            # if this is a retry, try to trim down the request to avoid ones we've already processed
+            self._request["rows"] = _ReadRowsOperation._revise_request_rowset(
+                row_set=self._request.get("rows", None),
+                last_seen_row_key=self._last_seen_row_key,
+            )
+            # revise next request's row limit based on number emitted
+            if total_row_limit:
+                new_limit = total_row_limit - self._emit_count
+                if new_limit <= 0:
+                    return
+                else:
+                    self._request["rows_limit"] = new_limit
+        new_gapic_stream = await gapic_fn(
+            self._request,
+            timeout=per_request_timeout,
+        )
+        buffer: asyncio.Queue[Row | RequestStats | Exception] = asyncio.Queue(
+            maxsize=buffer_size
+        )
+        buffer_task = asyncio.create_task(
+            self._generator_to_buffer(buffer, new_gapic_stream)
+        )
+        buffered_stream = self._buffer_to_generator(buffer)
+        state_machine = _StateMachine()
+        try:
+            stream = _ReadRowsOperation.merge_row_response_stream(
+                buffered_stream, state_machine
+            )
+            # run until we get a timeout or the stream is exhausted
+            while True:
+                new_item = await asyncio.wait_for(
+                    stream.__anext__(), timeout=per_row_timeout
+                )
```

Comment:
It looks like this `wait_for` method is very slow. I'll have to find a different way to handle the per_row_timeout, without the overhead of creating a new task. 

Or is per_row_timeout really even necessary? I recall that it works pretty differently in Java. Maybe we can just drop this and rely on the `per_attempt` and `operation` timeouts?

This can probably come in a future optimization PR, but let me know if you have thoughts

Reply:
Java specifically uses a global thread to avoid the perf penalty. I think you are right it might not be necessary in python. The primary reason it exists in java is because async programming in java is errorprone and forgetting to chain a callback leads to difficult to debug client hangs. The read timeout is a safe guard to ensure that if there was a bug somewhere, that nothing would ever hang. In python with async/await, I'm not as sure of its utility
We can drop it for now and re-add it in a follow up.